### PR TITLE
OHRM5X-1245: [BE] System Check

### PR DESCRIPTION
**Pull request by SandeepaOHRM**

### CODE REVIEW

File: installer/Util/Constant.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Installer\Util;
+
+final class Constant
```
Comment (RajithaKumara) [inline installer/Util/Constant.php:22]: we can remove this right. can you check usage

File: installer/Util/Messages.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Installer\Util;
+
+class Messages
```
Comment (RajithaKumara) [inline installer/Util/Messages.php:22]: remove non used messages

File: installer/Util/Messages.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    public const MYSQL_CLIENT_RECOMMEND_MESSAGE = "MySQL Client - ver 4.1.x or later recommended";
+    public const MYSQL_CLIENT_FAIL_MESSAGE = "MySQL Client - MySQL support not available in PHP settings";
+
+    public const MYSQL_SERVER_OK_MESSAGE = "MySQL Server - Ok";
+    public const MYSQL_SERVER_RECOMMEND_MESSAGE = "MySQL Server - ver 5.1.6 or later recommended";
+    public const MYSQL_SERVER_FAIL_MESSAGE = "MySQL Server - Not Available";
+
+    public const WEB_SERVER_OK_MESSAGE = "OK";
+
+    public const WritableLibConfs_OK_MESSAGE = "Write Permissions for lib/confs/ - Writeable";
```
Comment (RajithaKumara) [inline installer/Util/Messages.php:41]:
```suggestion
    public const WRITABLE_LIB_CONF_OK_MESSAGE = "Write Permissions for lib/confs/ - Writeable";
```

File: installer/Util/SystemConfig.php [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+    }
+
+    /**
+     * MYSQL Client Check
+     * @return array
+     */
+    public function isMySqlClientCompatible(): array
+    {
+        if (function_exists('mysqli_get_client_info')) {
+            $mysqlClient = mysqli_get_client_info();
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:95]: is there any way to get this via PDO

File: installer/Util/SystemConfig.php [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+     * MYSQL Server Check
+     * @return array
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function isMySqlServerCompatible(): array
+    {
+        if ($this->checkConnection()) {
+            $connection = $this->getConnection();
+            $result = $connection->executeQuery("SELECT VERSION()")->fetchAssociative();
+            $mysqlServerVersion = $result['VERSION()'];
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:132]:
```suggestion
            $result = $connection->executeQuery("SELECT VERSION() as version")->fetchAssociative();
            $mysqlServerVersion = $result['version'];
```
:thinking:

File: installer/Util/SystemConfig.php [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+        }
+    }
+
+    /**
+     * Write Permissions for “symfony/cache” Check
+     * @return array
+     */
+    public function isWritableSymfonyCache(): array
+    {
+        if ($this->checkWritePermission(realpath(__DIR__ . '/../../symfony/cache'))) {
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:305]:
```suggestion
        if ($this->checkWritePermission(Config::get(Config::CACHE_DIR))) {
```

File: installer/Util/SystemConfig.php [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+        }
+    }
+
+    /**
+     * Write Permissions for “symfony/log” Check
+     * @return array
+     */
+    public function isWritableSymfonyLog(): array
+    {
+        if ($this->checkWritePermission(realpath(__DIR__ . '/../../symfony/log'))) {
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:325]:
```suggestion
        if ($this->checkWritePermission(Config::get(Config::LOG_DIR))) {
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                            'label' => 'Register Global turned-off',
+                            'value' => $systemConfig->isRegisterGlobalsOff()
+                        ],
+                        [
+                            'label' => 'Memory Allocated for php script',
+                            'value' => $systemConfig->getAllocatedMemoryStatus()
+                        ],
+                        [
+                            'label' => 'Web Server allows .htacess files',
+                            'value' => []
```
Comment (SandeepaOHRM) [inline installer/Controller/Upgrader/Api/SystemCheckAPI.php:103]: This need to finalize with .htaccess

File: installer/Util/Constant.php (hunk 1 above)
Comment (SandeepaOHRM) [reply installer/Util/Constant.php:22]: done

File: installer/Util/SystemConfig.php (hunk 5 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:132]: done

File: installer/Util/Messages.php (hunk 2 above)
Comment (SandeepaOHRM) [reply installer/Util/Messages.php:22]: done

File: installer/Util/Messages.php (hunk 3 above)
Comment (SandeepaOHRM) [reply installer/Util/Messages.php:41]: done

File: installer/Util/SystemConfig.php (hunk 4 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:95]: PDO::getAttribute(PDO::ATTR_CLIENT_VERSION)
found this one. But it throws an exception

File: installer/Util/SystemConfig.php (hunk 7 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:325]: done

File: installer/Util/SystemConfig.php (hunk 6 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:305]: done

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+                    'category' => 'Permissions',
+                    'checks' => [
+                        [
+                            'label' => 'Write Permissions for “lib/confs”',
+                            'value' => $systemConfig->isWritableLibConfs()
+                        ],
+                        [
+                            'label' => 'Write Permissions for “lib/logs”',
+                            'value' => $systemConfig->isWritableLibLogs()
+                        ],
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Api/SystemCheckAPI.php:71]: I think we can remove this. will keep it for the moment

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                        [
+                            'label' => 'Write Permissions for “symfony/configs”',
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Api/SystemCheckAPI.php:73]:
```suggestion
                            'label' => 'Write Permissions for “symfony/config”',
```

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+                            'value' => $systemConfig->isWritableSymfonyLog()
+                        ],
+                    ]
+                ],
+                [
+                    'category' => 'Extensions',
+                    'checks' => [
+                        [
+                            'label' => 'Maximum Session idle time before timeout',
+                            'value' => $systemConfig->isMaximumSessionIdle()
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Api/SystemCheckAPI.php:91]: note this also, we can remove it later

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                        ],
+                        [
+                            'label' => 'Register Global turned-off',
+                            'value' => $systemConfig->isRegisterGlobalsOff()
+                        ],
+                        [
+                            'label' => 'Memory Allocated for php script',
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Api/SystemCheckAPI.php:98]:
```suggestion
                            'label' => 'Memory Allocated for PHP script',
```

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php [hunk 13]
```diff
@@ -19,47 +19,50 @@
...
-    ): bool {
-        $this->dbConnection = @new mysqli($host, $username, $password, $dbname, $port);
-        return !$this->dbConnection->connect_error;
+    public function checkDatabaseConnection(): bool
+    {
+        try {
+            $connection = $this->getConnection();
+            $connection->connect();
+            return true;
+        } catch (Exception $exception) {
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:37]: Log the exception,
```php
use OrangeHRM\Installer\Util\Logger;
// ...
Logger::getLogger()->error($e->getMessage());
Logger::getLogger()->error($e->getTraceAsString());
```
https://github.com/RajithaKumara/orangehrm/blob/OHRM5X-1254/installer/Controller/AbstractInstallerRestController.php#L78-L79

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php [hunk 14]
```diff
@@ -19,47 +19,50 @@
...
-        $result = $this->executeSql($query);
-        return $result->num_rows > 0;
+        $connection = $this->getConnection();
+        return $connection->createSchemaManager()->tablesExist(['ohrm_upgrade_status']);
+    }
+
+    /**
+     * @return \Doctrine\DBAL\Connection
+     */
+    public function getConnection(): \Doctrine\DBAL\Connection
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:55]:
```suggestion
    private function getConnection(): \Doctrine\DBAL\Connection
```
:thinking:

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php [hunk 15]
```diff
@@ -19,47 +19,50 @@
...
     }
 
     /**
-     * @param string $query
-     * @return mixed
+     * @return AbstractSchemaManager
+     * @throws \Doctrine\DBAL\Exception
      */
-    private function executeSql(string $query)
+    public function getSchemaManager(): AbstractSchemaManager
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:64]:
```suggestion
    private function getSchemaManager(): AbstractSchemaManager
```

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php [hunk 16]
```diff
@@ -19,47 +19,50 @@
...
      * @return bool
+     * @throws \Doctrine\DBAL\Exception
      */
     public function checkDatabaseStatus(): bool
     {
-        $query = "SHOW TABLES LIKE 'ohrm_upgrade_status'";
-        $result = $this->executeSql($query);
-        return $result->num_rows > 0;
+        $connection = $this->getConnection();
+        return $connection->createSchemaManager()->tablesExist(['ohrm_upgrade_status']);
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:49]:
```suggestion
        return $this->getSchemaManager()->tablesExist(['ohrm_upgrade_status']);
```

File: installer/Util/Messages.php [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    public const CURLStatus_OK_MESSAGE = "cURL status - Enabled";
+
+    public const SimpleXMLStatus_DISABLE_MESSAGE = "SimpleXML status - Disabled. SimpleXML, libxml and xml PHP libraries are required";
+    public const SimpleXMLStatus_OK_MESSAGE = "SimpleXML status - Enabled";
+
+    public const ZIP_Status_DISABLE_MESSAGE = "ZIP status - Disabled. SimpleXML, libxml and xml PHP libraries are required";
+    public const ZIP_Status_OK_MESSAGE = "ZIP status - Enabled";
+
+    public const PHP_MIN_VERSION = '5.3.0';
+    public const MYSQL_MIN_VERSION = '4.1.0';
```
Comment (RajithaKumara) [inline installer/Util/Messages.php:75]: Seems wrong

File: installer/Util/Messages.php [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+ */
+
+namespace OrangeHRM\Installer\Util;
+
+class Messages
+{
+    public const SEPERATOR = "----------------------------------";
+
+    public const PHP_OK_MESSAGE = "PHP Version - Ok";
+    public const PHP_FAIL_MESSAGE = "PHP Version - PHP 5.3.0 or higher is required";
```
Comment (RajithaKumara) [inline installer/Util/Messages.php:27]: BE we have version validation, https://github.com/orangehrm/orangehrm/blob/5.0/installer/environmentCheck/system_requirements.yml

may have to generate messages dynamically. I'll highlight in # same places

File: installer/Util/Messages.php [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+    public const MYSQL_CLIENT_OK_MESSAGE = "MySQL Client - Ok";
+    public const MYSQL_CLIENT_RECOMMEND_MESSAGE = "MySQL Client - ver 4.1.x or later recommended";
```
Comment (RajithaKumara) [inline installer/Util/Messages.php:30]: \#

File: installer/Util/Messages.php [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    public const MYSQL_CLIENT_FAIL_MESSAGE = "MySQL Client - MySQL support not available in PHP settings";
+
+    public const MYSQL_SERVER_OK_MESSAGE = "MySQL Server - Ok";
+    public const MYSQL_SERVER_RECOMMEND_MESSAGE = "MySQL Server - ver 5.1.6 or later recommended";
```
Comment (RajithaKumara) [inline installer/Util/Messages.php:34]: \#

File: installer/Util/Messages.php [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+    public const RegisterGlobalsOff_OK_MESSAGE = "Ok";
+    public const RegisterGlobalsOff_FAIL_MESSAGE = "On. Should be off";
+
+    public const GgExtensionEnable_OK_MESSAGE = "PHP gd extension - Enabled";
+    public const GgExtensionEnable_FAIL_MESSAGE = "PHP gd extension - Not enabled";
+
+    public const MySQLEventStatus_FAIL_MESSAGE = "MySQL Event Scheduler status - Cannot connect to the database";
+    public const MySQLEventStatus_DISABLE_MESSAGE = "MySQL Event Scheduler status - Disabled. This is required for automatic leave status changes of Leave module";
+    public const MySQLEventStatus_OK_MESSAGE = "MySQL Event Scheduler status - Enabled";
```
Comment (RajithaKumara) [inline installer/Util/Messages.php:63]:
```suggestion
```

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php [hunk 22]
```diff
@@ -19,47 +19,50 @@
 
 namespace OrangeHRM\Installer\Controller\Upgrader\Traits;
 
-use mysqli;
+use Doctrine\DBAL\Schema\AbstractSchemaManager;
+use Exception;
+use OrangeHRM\Installer\Util\Connection;
 
 trait UpgraderUtilityTrait
```
Comment (RajithaKumara) [inline installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:26]: I think this should be a class, not a trait

File: installer/Util/SystemConfig.php [hunk 23]
```diff
@@ -0,0 +1,622 @@
...
+     * @return array
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function isMySqlServerCompatible(): array
+    {
+        if ($this->checkConnection()) {
+            $connection = $this->getConnection();
+            $result = $connection->executeQuery("SELECT VERSION() AS version")->fetchAssociative();
+            $mysqlServerVersion = $result['version'];
+            if (version_compare($mysqlServerVersion, "5.1.6") >= 0) {
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:134]: \#

File: installer/Util/SystemConfig.php [hunk 24]
```diff
@@ -0,0 +1,622 @@
...
+        }
+    }
+
+    /**
+     * Web Server Check
+     * @return array
+     */
+    public function isWebServerCompatible(): array
+    {
+        $supportedWebServers = ['Apache', 'nginx', 'IIS'];
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:219]: https://github.com/orangehrm/orangehrm/blob/5.0/installer/environmentCheck/system_requirements.yml#L16

File: installer/Util/SystemConfig.php [hunk 25]
```diff
@@ -0,0 +1,622 @@
...
+        }
+    }
+
+    /**
+     * Write Permissions for “symfony/configs” Check
+     * @return array
+     */
+    public function isWritableSymfonyConfig(): array
+    {
+        if ($this->checkWritePermission(realpath(__DIR__ . '/../../symfony/config'))) {
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:286]: better to define `Config::CONFIG_DIR`

File: installer/Util/SystemConfig.php [hunk 26]
```diff
@@ -0,0 +1,622 @@
...
+        return [
+            'message' => "Memory allocated for PHP script - ${message}",
+            'status' => $status
+        ];
+    }
+
+    /**
+     * @return array
+     */
+    public function IsGgExtensionEnable(): array
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:437]: are we using this?? :open_mouth:

File: installer/Util/SystemConfig.php [hunk 27]
```diff
@@ -0,0 +1,622 @@
...
+                'message' => Messages::ZIP_Status_DISABLE_MESSAGE,
+                'status' => self::BLOCKER
+            ];
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    private function checkConnection(): bool
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:548]: saw this was there in the trait as well

File: installer/Util/SystemConfig.php [hunk 28]
```diff
@@ -0,0 +1,622 @@
...
+     * @return bool
+     */
+    private function checkWritePermission(string $path): bool
+    {
+        try {
+            $this->filesystem->dumpFile($path . '/_temp.txt', $path);
+            $this->filesystem->remove($path . '/_temp.txt');
+            return true;
+        } catch (Exception $e) {
+            return false;
```
Comment (RajithaKumara) [inline installer/Util/SystemConfig.php:572]: log exception

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Review by RajithaKumara — APPROVED** [review]

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php (hunk 11 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Api/SystemCheckAPI.php:91]: okay

File: installer/Util/SystemConfig.php (hunk 27 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:548]: removed

File: installer/Util/SystemConfig.php (hunk 28 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:572]: done

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php (hunk 8 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Api/SystemCheckAPI.php]: removed

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php (hunk 10 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Api/SystemCheckAPI.php:73]: done

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php (hunk 13 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:37]: done

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php (hunk 16 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:49]: done

File: installer/Util/Messages.php (hunk 17 above)
Comment (SandeepaOHRM) [reply installer/Util/Messages.php:75]: updated to dynamic versions

File: installer/Util/Messages.php (hunk 18 above)
Comment (SandeepaOHRM) [reply installer/Util/Messages.php:27]: done

File: installer/Util/Messages.php (hunk 19 above)
Comment (SandeepaOHRM) [reply installer/Util/Messages.php:30]: done

File: installer/Util/Messages.php (hunk 20 above)
Comment (SandeepaOHRM) [reply installer/Util/Messages.php:34]: done

File: installer/Util/Messages.php (hunk 21 above)
Comment (SandeepaOHRM) [reply installer/Util/Messages.php:63]: removed

File: installer/Util/SystemConfig.php (hunk 23 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:134]: removed

File: installer/Util/SystemConfig.php (hunk 24 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:219]: modified

File: installer/Util/SystemConfig.php (hunk 26 above)
Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:437]: removed

Comment (SandeepaOHRM) [reply installer/Util/SystemConfig.php:437]: removed

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php (hunk 22 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:26]: created a utility class

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php (hunk 15 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:64]: done

File: installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php (hunk 14 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Traits/UpgraderUtilityTrait.php:55]: This is also used in the SystemConfig class

File: installer/Controller/Upgrader/Api/SystemCheckAPI.php (hunk 12 above)
Comment (SandeepaOHRM) [reply installer/Controller/Upgrader/Api/SystemCheckAPI.php:98]: done

**Pull request merged**